### PR TITLE
wdio-allure-reporter: Fixes #11473

### LINKED ISSUE #11473: [🐛 Bug]: wdio-allure-reporter: Inaccurate Reporting and Statistics for Hooks as Test Cases
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md#reporting-new-issues).

### WebdriverIO Version

latest

### Node.js Version

latest

### Mode

WDIO Testrunner

### What happened?

The current issue arises from the inability to utilize the fixtures provided by the AllureJS project effectively. Currently, hooks are being reported as tests, leading to several significant problems with the generated reports and statistics.

**Misleading Reports:** The primary issue is that hooks, which are not actual test cases, are being incorrectly reported as tests. This misrepresentation creates confusion and ambiguity in the generated reports.

**Statistics Distortion:** Reporting hooks as tests distorts the statistics related to the number of actual test cases. This affects the accuracy of execution statistics, making it challenging to maintain clear and reliable execution data.

**Execution Uniqueness:** The inaccurate reporting of hooks as tests can lead to difficulties in distinguishing between the execution of actual test cases and the execution of setup/teardown hooks. This can hinder the ability to track the uniqueness of test executions accurately.

**Currently:**
image

### What is your expected behavior?

The expected behavior, once this issue is resolved, should include the following:

**Accurate Reporting:** Hooks will be correctly identified as fixtures, not as test cases, in the AllureJS project. This will ensure that the generated reports accurately reflect the nature of the executed code and tests.

**Clear Statistics:** With the accurate identification of hooks as fixtures, the statistics related to test cases will no longer be distorted. This should lead to clear and reliable statistics about the execution of actual test cases.

**Distinct Execution Tracking:** The resolution of this issue should enable a clear distinction between the execution of actual test cases and the execution of setup/teardown hooks. This will help in maintaining the uniqueness and reliability of test execution history.

Overall, the expected behavior is that the changes and improvements proposed in this issue will result in a more precise, reliable, and comprehensible reports.

**Example:**
image

### How to reproduce the bug.

To reproduce this bug, simply create a new project with Allure, add a test case with hooks, and generate an Allure report using the command line.

### Relevant log output

```typescript
N/A
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Is there an existing issue for this?

- [X] I have searched the existing issues

### PR DESCRIPTION
**Pull request by lararojasmr**
## Proposed changes

I have made an improvement in the Allure reporter in which I have adapted the methods to support fixtures with the goal of enhancing the report, preserving statistics, and keeping hooks within the test cases. Testing has been conducted, and test cases have been adjusted to accommodate the new functionality.

Furthermore, I have changed the 'disableMochaHooks' option to 'disableHooks' since it now also supports Jasmine. Additionally, I have introduced a new option: 'testStatusFailedHook,' which allows specifying the default state for a hook when an explicit state is not provided. By default, this state is set to 'Skipped.'

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

**Comment by BorisOsipov** [comment]
@lararojasmr As a first step, undo all formatting changes and make eslint happy.

**Comment by lararojasmr** [comment]
> @lararojasmr As a first step, undo all formatting changes and make eslint happy.

I've executed the project eslint with fix haha. Can you confirm the command that I need to execute?

**Comment by BorisOsipov** [comment]
@lararojasmr You can peek here https://github.com/webdriverio/webdriverio/blob/main/package.json#L38

**Comment by lararojasmr** [comment]
> Member

Yes, I executed this command: `npm run test:eslint -- --fix`

File: packages/wdio-allure-reporter/src/reporter.ts
Comment (BorisOsipov) [inline packages/wdio-allure-reporter/src/reporter.ts:636]: The error object must pass through `getErrorFromFailedTest(err)' to properly format errors from different test runners. e.g. :

```
const error = getErrorFromFailedTest(isFailed)
entity.detailsMessage = error.message
entity.detailsTrace = error.stack
```

File: packages/wdio-allure-reporter/src/reporter.ts
Comment (BorisOsipov) [inline packages/wdio-allure-reporter/src/reporter.ts:681]: The error object must pass through `getErrorFromFailedTest(err)' to properly format errors from different test runners.

**Comment by BorisOsipov** [comment]
I did a quick test with Jasmine and ran into a problem

```
import {expect} from 'expect-webdriverio'
describe('foo', () => {
    //same if we use afterAll
    beforeAll(() => {
        expect(true).toEqual(false)
    });
    it('t1', () => {

    })

    it('t2', () => {

    })
})
```

There are 3 tests, but there should be 2

image

**Comment by lararojasmr** [comment]
> I did a quick test with Jasmine and ran into a problem
> 
> ```
> import {expect} from 'expect-webdriverio'
> describe('foo', () => {
>     //same if we use afterAll
>     beforeAll(() => {
>         expect(true).toEqual(false)
>     });
>     it('t1', () => {
> 
>     })
> 
>     it('t2', () => {
> 
>     })
> })
> ```
> 
> There are 3 tests, but there should be 2
> 
> image

Hey @BorisOsipov I'm was able to reproduce that issue:

Using your test case I got:

image
**Please remove:** `import {expect} from 'expect-webdriverio'`

**Example (beforeAll)**
image

**Example(afterAll)**
image

I made in order to fix the issue, good catch, thanks!

File: packages/wdio-allure-reporter/tests/cucumber.suite.test.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/tests/cucumber.suite.test.ts:4]:
```suggestion
import {
  Status,
  LabelName,
  LinkType,
  Stage,
  type Label,
  type Parameter,
  type Link,
  type StepResult
} from 'allure-js-commons'
```

File: packages/wdio-allure-reporter/src/state.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/state.ts:4]: Why do all these imports have this weird formatting?

File: packages/wdio-allure-reporter/tests/suite.test.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/tests/suite.test.ts:5]: same

File: packages/wdio-allure-reporter/src/state.ts
Comment (lararojasmr) [reply packages/wdio-allure-reporter/src/state.ts]: intelliJIDEA  😆 I've fixed it

**Comment by lararojasmr** [comment]
I'm fixing a use case that fails.

**Comment by lararojasmr** [comment]
@BorisOsipov @christian-bromann. Now the code was updated and include all fixes. I ran 394 test cases and fix all issues that I found.

File: packages/wdio-allure-reporter/tests/allureFeatures.test.ts
Comment (christian-bromann) [inline packages/wdio-allure-reporter/tests/allureFeatures.test.ts:1186]:
```suggestion
    it('should create a test and add mocha all hooks as fixtures if hook throws', () => {
```

**Review by BorisOsipov — CHANGES_REQUESTED** [review]
Please remove all unnecessary, unrelated reformatting changes from the diff. It pollutes git blame and makes problems difficult to investigate.

**Review by BorisOsipov — CHANGES_REQUESTED** [review]
Please remove all unnecessary, unrelated reformatting changes from the diff. It pollutes git blame and makes problems difficult to investigate.

**Comment by BorisOsipov** [comment]
Damn, Github didn't publish my comments on the review in time.I It is ridiculous to make review with tool that publishes few comments immediately and not publish all other until I click request changes button 😞

File: packages/wdio-allure-reporter/README.md
Comment (lararojasmr) [reply packages/wdio-allure-reporter/README.md:53]: ok, w can put this change in other ticket. I'll work on this now

Comment (BorisOsipov) [reply packages/wdio-allure-reporter/README.md:53]: @lararojasmr Yes, create a ticket to rename it and add a comment with it in https://github.com/webdriverio/webdriverio/issues/10362

File: packages/wdio-allure-reporter/src/reporter.ts
Comment (lararojasmr) [reply packages/wdio-allure-reporter/src/reporter.ts]: This method is pretty basic, does not take into account steps result values, only cover step assertion or not, if not assertion the result is broken by default. Let me work on this method to increase the capabilities.

Comment (BorisOsipov) [reply packages/wdio-allure-reporter/src/reporter.ts]: @lararojasmr You don't need to do any extra work here. This is the cross-language and cross-framework behavior of allure. Status is "failed" if there is an assertion error, and "broken" in all other cases.

File: packages/wdio-allure-reporter/src/reporter.ts
Comment (lararojasmr) [reply packages/wdio-allure-reporter/src/reporter.ts]: In the fixtures, expect does not work as same test cases. So if a hook does not throw an error, the hook will be mark as passed. I think that this method was implemented thinking that both are test cases or steps.

File: packages/wdio-allure-reporter/src/reporter.ts
Comment (lararojasmr) [reply packages/wdio-allure-reporter/src/reporter.ts]: yes, comment is misleading I've changed that.

File: packages/wdio-allure-reporter/src/reporter.ts
Comment (lararojasmr) [reply packages/wdio-allure-reporter/src/reporter.ts:681]: Fixed

File: packages/wdio-allure-reporter/src/reporter.ts
Comment (lararojasmr) [reply packages/wdio-allure-reporter/src/reporter.ts:682]: fixed

File: packages/wdio-allure-reporter/src/reporter.ts
Comment (lararojasmr) [reply packages/wdio-allure-reporter/src/reporter.ts]: Ok, I'm working on a new changes for this method, hooks (fixtures), test and steps are different way to setup the status.

**Comment by lararojasmr** [comment]
Thanks for the review, today or tomorrow I'll send my new changes covering all suggestions.

**Comment by lararojasmr** [comment]
Hey team,  I made a new commit to PR with all changes, i ran tests, check the only and the import (the format was a IDE config sorry).

File: packages/wdio-allure-reporter/src/types.ts
Comment (lararojasmr) [reply packages/wdio-allure-reporter/src/types.ts:54]: working on it now.

File: packages/wdio-allure-reporter/README.md
Comment (lararojasmr) [reply packages/wdio-allure-reporter/README.md:53]: I set the variable name in: **disableMochaHooks** but I left the new description.

File: packages/wdio-allure-reporter/src/types.ts
Comment (lararojasmr) [reply packages/wdio-allure-reporter/src/types.ts:54]: @BorisOsipov I'll put off this option for next ticket with renaming of the option for mocha as next improvements

File: packages/wdio-allure-reporter/src/reporter.ts
Comment (lararojasmr) [reply packages/wdio-allure-reporter/src/reporter.ts]: But, fixtures do not work in the same way, because, if a hook return error undefined and non status, the result is not broken is passed. So the idea is validate when a step is related to hook or not so into a fixture the hook is an step as same as test. With that solution all fixtures get a Broken result.

Comment (lararojasmr) [reply packages/wdio-allure-reporter/src/reporter.ts]: I've implemented an idea that i have, take a look and give me your comments.

**Comment by lararojasmr** [comment]
@christian-bromann i need to do something about this error: 
image

**Comment by christian-bromann** [comment]
> do I need to do something about this error?:

no, all good!

**Comment by BorisOsipov** [comment]
@lararojasmr It is a busy week, I will find time to review this in the next few days.

**Comment by christian-bromann** [comment]
@lararojasmr there seem to be some type errors from the Allure reporter, mind taking a look?

**Comment by lararojasmr** [comment]
@christian-bromann @BorisOsipov I was sick in the days before, so now I'm continuing to work on this.

**Comment by lararojasmr** [comment]
@christian-bromann can you run the checks again? I updated branch and pull request. I think that all is running as expected for fixtures and tests.

**Comment by lararojasmr** [comment]
@christian-bromann @BorisOsipov 
image
???

**Comment by christian-bromann** [comment]
@lararojasmr don't worry about this failure, Chrome is starting to phase out `userAgent` which we use in some of our e2e tests. I am happy to merge this once @BorisOsipov approves.

**Comment by christian-bromann** [comment]
Mentioned bug has been resolved and build is expected to pass now. Still waiting for @BorisOsipov to approve.

**Comment by BorisOsipov** [comment]
I will find time to look into this PR tomorrow.

**Comment by lararojasmr** [comment]
@christian-bromann image

**Comment by BorisOsipov** [comment]
Issue #1
```
//mocha

describe('foo', () => {
    it('t1', async () => {
        await browser.url("https://www.york.ac.uk/")
    })

    describe('foo2', () => {
        beforeEach(async () => {
            expect(true).toEqual(false)
        });
        it('t1', async () => {
            await browser.url("https://www.york.ac.uk/")
        })
    })
})
```
Produces broken report https://imgur.com/a/E79hty2

**Comment by BorisOsipov** [comment]
Issue 2
```
//mocha

describe('foo', () => {
    beforeEach(async () => {
        expect(true).toEqual(false)
    });

    it('t1', async () => {
    })
})
```
Produces broken report https://imgur.com/a/3uMlTje

**Comment by BorisOsipov** [comment]
Issue 3
```
//mocha
describe('foo', () => {
    before(async () => {
        expect(true).toEqual(false)
    });

    it('t1', async () => {
    })
})
```
Produces broken report https://imgur.com/a/raEDlfQ - The suite name is <Empty>

**Comment by BorisOsipov** [comment]
Issue 0

Please.I have written this at least twice, as I can see. Please remove all unnecessary, unrelated reformatting changes from the diff. It pollutes git blame and makes problems difficult to investigate.
I cannot accept these changes, they are harmful.

**Review by BorisOsipov — CHANGES_REQUESTED** [review]
See comments.

File: packages/wdio-allure-reporter/tests/utils.test.ts
Comment (lararojasmr) [reply packages/wdio-allure-reporter/tests/utils.test.ts:44]: But this logic is only return FAILED. Could you explain me why we are doing this here? right now, jasmine and mocha return a similar structure, why we cannot consider the error fields instead?

Comment (BorisOsipov) [reply packages/wdio-allure-reporter/tests/utils.test.ts:44]: @lararojasmr  It seems I did not explain well enough
> why we cannot consider the error fields instead?

We can, and this is a good substitute of config check solution. Please add test to branch
```
} else if (test.errors) {
        return AllureStatus.FAILED
    }
```

Comment (BorisOsipov) [reply packages/wdio-allure-reporter/tests/utils.test.ts:44]: >right now, jasmine and mocha return a similar structure, why we cannot consider the error fields instead?

Jasmine has soft assertions that can return multiple errors and we can't properly calculate the status, so we always set it as failed.

File: packages/wdio-allure-reporter/src/utils.ts
Comment (lararojasmr) [reply packages/wdio-allure-reporter/src/utils.ts:142]: @BorisOsipov use expect does not work fine in **All** hooks, why are you using it in that places? I saw that hooks does not recognize the status and errors using assertion in that places. I think that is a limitation with (All) fixtures.

The method **getTestStatus**, is only covering test cases scope. And the hook as fixtures has a different behavior, please check the object response.

Each Hooks are inside a test case scope so it will be affect the hook object status and error.

The behavior is expected in your test case, use throw error instead to see a similar behavior. I check this with a three different projects, cucumber, mocha, jasmine and mi company project (use mocha)

File: packages/wdio-allure-reporter/tests/utils.test.ts
Comment (lararojasmr) [reply packages/wdio-allure-reporter/tests/utils.test.ts:44]: @BorisOsipov What is the logic for jasmine that you are seeing? I removed that logic (from getTestStatus) because after testing both frameworks i saw that the behavior is similar and it does not need specific implementation for the actions that the allure service do.

Comment (BorisOsipov) [reply packages/wdio-allure-reporter/tests/utils.test.ts:44]: >What is the logic for jasmine that you are seeing? I removed that logic (from getTestStatus) because after testing both frameworks i saw that the behavior is similar and it does not need specific implementation for the actions that the allure service do.

It depends which asserts you use, jasmine or wdio-expect. I'd prefer not to change the logic, at least for this PR. I I don't want to test it now with all the combimation due, I'm not a Jasmine user :)

`test.errors` value sets only for Jasmine integration. It always empty for mocha\cucumber. And I want to see that when we use Jasmine we always have FAILED test status. And I want to see test that when `test.errors` is set to status always failed to maintain test coverage.

Comment (BorisOsipov) [reply packages/wdio-allure-reporter/tests/utils.test.ts:44]: It is just a matter of adding a unit test for this change.

File: packages/wdio-allure-reporter/src/utils.ts
Comment (lararojasmr) [reply packages/wdio-allure-reporter/src/utils.ts:142]: Ok, @BorisOsipov AssertionError won't be recognized in 'all' type fixtures. This is because you can use it, but the hook object that leads to the methods is not affected in any way, and the error is not logged as it's not within the context of a test case yet. Previously, the service considered all fixtures as steps within the context of a test case or as a test case within the context of a suite, and your solution worked, but now it doesn't work that way.

Comment (lararojasmr) [reply packages/wdio-allure-reporter/src/utils.ts:142]: In the code below the hook will be broken, and the test case will be broken, because the **beforeAll** fixture is in the suite scope not test, and the test does not have a assertion error. So the test should be BROKEN.
```
describe('foo', () => {
    beforeAll(async () => {
        expect(true).toEqual(false)
    });
    it('t1', async () => {
    })
})
```

In the code below the test case will be failed, because the **beforeEach** fixture is in the test (t1) scope, and the test it will be affected by the fixture assertion. So the test should be FAILED. 
```
describe('foo', () => {
    beforeEach(async () => {
        expect(true).toEqual(false)
    });
    it('t1', async () => {
    })
})
```

It's the way are fixtures works not my way. I read all, and I debug the objects with different implementations, I saw allure-js-commons code, and you are asking me to change fixtures for test/steps again.

Comment (lararojasmr) [reply packages/wdio-allure-reporter/src/utils.ts:142]: @christian-bromann could you give us your opinion on this? I think that we are having a misunderstanding.

Comment (BorisOsipov) [reply packages/wdio-allure-reporter/src/utils.ts:142]: @lararojasmr You seem to have more experience with the actual implementation details. Could you provide an example of test project code where the following `getHookStatus` simplification causes problems? I mean sample code and screen like I do in https://github.com/webdriverio/webdriverio/pull/11475#issuecomment-1791399930

```
export const getHookStatus = (newHookStats: HookStats, hookElement: ExecutableItemWrapper, hookRootStep: AllureStep) => {
    hookElement.stage = hookRootStep.stage = Stage.FINISHED
    const formattedError = getErrorFromFailedTest(newHookStats)
    hookElement.detailsMessage = hookRootStep.detailsMessage = formattedError?.message
    hookElement.detailsTrace = hookRootStep.detailsTrace = formattedError?.stack
    const finalStatus = getTestStatus(newHookStats)
    hookElement.status = hookRootStep.status = finalStatus
}
```

I find it extremely difficult to understand this problem without an example. Please explain with example.

Comment (BorisOsipov) [reply packages/wdio-allure-reporter/src/utils.ts:142]: >I think that we are having a misunderstanding.

Yeap. Now I get want you want to explain. Let me think.

**Comment by christian-bromann** [comment]
@lararojasmr anything I can do to help bring this over the goal line?

**Comment by lararojasmr** [comment]
> @lararojasmr anything I can do to help bring this over the goal line?

Hey @christian-bromann i was very busy last week but i hope work on this today. I just need to create a project to show to @BorisOsipov how fixtures are processed by allure, create some test and change some formatting things. 

I hope to finish with that today.

**Comment by christian-bromann** [comment]
@lararojasmr let me know if I can help with anything.

**Comment by christian-bromann** [comment]
@lararojasmr @BorisOsipov let me know if I can help somehow to push this over the goal line.

**Comment by lararojasmr** [comment]
@christian-bromann @BorisOsipov hey team, sorry my delay but I hope to finish with this after I fix and issue that i have with appium and webdriverio with the latest versions, the test case is not starting and is stuck in spawn but without timeout. But when i solved that i'll finish with this code that also i need. Sorry again.

**Pull request closed without merging**
